### PR TITLE
Measure and expose DNS programming latency from Kubernetes plugin.

### PR DESCRIPTION
**Pull request by oxddr**
For now metric is measure only for headless services. Informer has been slighlty
refactored, so the code can measure latency without storing extra fields on
Endpoint struct.

### 1. Why is this pull request needed and what does it do?
Measure and expose DNS programming latency from Kubernetes plugin.

### 2. Which issues (if any) are related?
ref https://github.com/kubernetes/perf-tests/issues/617

### 3. Which documentation changes (if any) need to be made?
none 

### 4. Does this introduce a backward incompatible change or deprecation?
no

### CODE REVIEW

**Comment by oxddr** [comment]
@johnbelamaric any chance you can have a look at this PR? It follows your suggestions on how measuring should be done without increasing memory footprint.

File: plugin/kubernetes/object/informer.go [hunk 1]
```diff
@@ -5,19 +5,25 @@ import (
...
-func NewIndexerInformer(lw cache.ListerWatcher, objType runtime.Object, h cache.ResourceEventHandler, indexers cache.Indexers, convert ToFunc) (cache.Indexer, cache.Controller) {
+// NewIndexerInformer is a copy of the cache.NewIndexerInformer function, but allows custom process function
+func NewIndexerInformer(lw cache.ListerWatcher, objType runtime.Object, h cache.ResourceEventHandler, indexers cache.Indexers, builder ProcessorBuilder) (cache.Indexer, cache.Controller) {
 	clientState := cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc, indexers)
 
-	fifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, clientState)
-
 	cfg := &cache.Config{
-		Queue:            fifo,
+		Queue:            cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, clientState),
```
Comment (johnbelamaric) [inline plugin/kubernetes/object/informer.go:13]: in general please do this sort of cleanup in a separate PR to reduce the review burden

File: plugin/kubernetes/metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+const (
+	namespace = "testns"
+)
+
+func TestDnsProgrammingLatency(t *testing.T) {
+	client := fake.NewSimpleClientset()
+	now := time.Now()
+	controller := newdnsController(client, dnsControlOpts{
+		initEndpointsCache: true,
+		// This is needed as otherwise the fake k8s client doesn't work properly.
```
Comment (johnbelamaric) [inline plugin/kubernetes/metrics_test.go:27]: Can you explain this more? If we can eliminate `skipAPIObjectsCleanup` this all becomes a lot simpler. I don't remember why we need that as an option or see anywhere it gets set?

File: plugin/kubernetes/object/informer.go (hunk 1 above)
Comment (oxddr) [reply plugin/kubernetes/object/informer.go:13]: Sure thing, I can remove it. This one simply felt like a trivial change.

Comment (johnbelamaric) [reply plugin/kubernetes/object/informer.go:13]: Not necessary, was just a note for the future.

File: plugin/kubernetes/metrics_test.go (hunk 2 above)
Comment (oxddr) [reply plugin/kubernetes/metrics_test.go:27]: This is needed purely for testing purposes. If we do cleanup resources in test, cache doesn't work properly and tests fail. I am not at my workstation at the moment, so I can't check the exact effect. 

One option is to remove tests. However this non-trivial change to the code and I'd rather have it tested.

File: plugin/kubernetes/object/informer.go (hunk 1 above)
Comment (oxddr) [reply plugin/kubernetes/object/informer.go:13]: Noted :)

File: plugin/kubernetes/metrics.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	return val
+}
+
+// isEndpointForHeadlessService returns true whether the endpoints object belongs to a headless
+// service (i.e. clusterIp = None). Note that this method may return false negatives if the service
+// informer is lagging, i.e. we may not see a recently created service. Given that the services
+// don't change very often (comparing to much more frequent endpoints changes), cases when this method
+// will return wrong answer should be relatively rare. Because of that we intentionally accept this
+// flaw to keep the solution simple.
+func isEndpointForHeadlessService(svcs []*object.Service, endpoints *api.Endpoints) bool {
```
Comment (johnbelamaric) [inline plugin/kubernetes/metrics.go]: This is only used once. Instead just inline it (along with the comment)

File: plugin/kubernetes/metrics.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	// LastChangeTriggerTime annotation is set). It means that the corresponding service is a
+	// "headless service with selector".
+	DnsProgrammingLatency.WithLabelValues("headless_with_selector").
+		Observe(durationSinceFunc(lastChangeTriggerTime).Seconds())
+
+}
+
+// getLastChangeTriggerTime returns the time.Time value of the EndpointsLastChangeTriggerTime
+// annotation stored in the given endpoints object or the "zero" time if the annotation wasn't set
+func getLastChangeTriggerTime(endpoints *api.Endpoints) time.Time {
```
Comment (johnbelamaric) [inline plugin/kubernetes/metrics.go]: this is only used once, instead just inline it along with the comment

File: plugin/kubernetes/metrics.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	// durationSinceFunc returns the duration elapsed since the given time.
+	// Added as a global variable to allow injection for testing.
+	durationSinceFunc = time.Since
+)
+
+func recordDNSProgrammingLatency(svcs []*object.Service, endpoints *api.Endpoints) {
+	lastChangeTriggerTime := getLastChangeTriggerTime(endpoints)
+
+	if endpoints == nil || !isEndpointForHeadlessService(svcs, endpoints) || lastChangeTriggerTime.IsZero() {
+		return
```
Comment (johnbelamaric) [inline plugin/kubernetes/metrics.go:66]: so we're not recording the latency for programming clusterIP or headless_without_selector? For clusterIP, I would think it's the service creation time vs. now - since clusterIP is immutable I think that should work.

For `headless_without_selector`, maybe the endpoints creation time...or update time if we have those in metadata. I recall you saying that we need the annotation though I can recall the exact reason offhand.

File: plugin/kubernetes/metrics.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+func getLastChangeTriggerTime(endpoints *api.Endpoints) time.Time {
+	stringVal, ok := endpoints.Annotations[api.EndpointsLastChangeTriggerTime]
+	if !ok {
+		// It's possible that the Endpoints object won't have the EndpointsLastChangeTriggerTime
+		// annotation set. In that case return the 'zero value', which is ignored in the upstream code.
+		return time.Time{}
+	}
+	val, err := time.Parse(time.RFC3339Nano, stringVal)
+	if err != nil {
+		log.Warningf("Error while parsing EndpointsLastChangeTriggerTimeAnnotation: '%s'. Error is %v",
```
Comment (johnbelamaric) [inline plugin/kubernetes/metrics.go]: Could this flood the logs?

If we're going to be wordy we should include enough info to debug and explain the actual effect we are warning about. The error value is not going to be useful I don't think, but knowing what it's supposed to look like will help.

```
log.Warningf("DnsProgrammingLatency cannot be calculated for Endpoints '%s/%s'; invalid %q annotation RFC3339 value of %q",
  endpoints.GetNamespace(), endpoints.GetName(), api.EndpointsLastChangeTriggerTime, stringVal)
```

**Review by johnbelamaric — CHANGES_REQUESTED** [review]

File: plugin/kubernetes/metrics.go (hunk 6 above)
Comment (oxddr) [reply plugin/kubernetes/metrics.go]: Thanks for the snippet!

I doubt. This annotations is machine-generated and it's not coming from user. I'd assume it should be formatted correctly unless we have a bug (should be easy to catch in tests) or a pathological case (user tampering with it).

File: plugin/kubernetes/metrics.go (hunk 5 above)
Comment (oxddr) [reply plugin/kubernetes/metrics.go:66]: Yes, this one does work only for headless services as for now. I was planning to follow up with the rest once this PR is hopefully merged. I assume adding those two missing types should be relatively easy.

**Comment by oxddr** [comment]
@johnbelamaric PTAL :)

**Comment by johnbelamaric** [comment]
I think we're OK here but I'd like to have @chrisohaver take a look as well.

/assign: @chrisohaver 
/lgtm

File: plugin/kubernetes/metrics_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
+package kubernetes
+
+import (
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coredns/coredns/plugin/kubernetes/object"
```
Comment (miekg) [inline plugin/kubernetes/metrics_test.go:8]: needs a newline after our local imports.

**Review by miekg — COMMENTED** [review]
this misses an explanation of the new metrics in the README.md

File: plugin/kubernetes/object/object.go [hunk 8]
```diff
@@ -16,14 +16,18 @@
 package object
 
 import (
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
+        "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (chrisohaver) [inline plugin/kubernetes/object/object.go]: maybe run a `gofmt` to clean it up this kind of thing...

File: plugin/kubernetes/object/service.go [hunk 9]
```diff
@@ -27,7 +27,13 @@ type Service struct {
 func ServiceKey(name, namespace string) string { return name + "." + namespace }
 
 // ToService converts an api.Service to a *Service.
```
Comment (chrisohaver) [inline plugin/kubernetes/object/service.go]:
```suggestion
// ToService returns a function that converts an api.Service to a *Service.
```

File: plugin/kubernetes/object/pod.go [hunk 10]
```diff
@@ -17,7 +17,13 @@ type Pod struct {
 }
 
 // ToPod converts an api.Pod to a *Pod.
```
Comment (chrisohaver) [inline plugin/kubernetes/object/pod.go]:
```suggestion
// ToPod returns a function that converts an api.Pod to a *Pod.
```

**Review by chrisohaver — APPROVED** [review]

File: plugin/kubernetes/object/object.go (hunk 8 above)
Comment (oxddr) [reply plugin/kubernetes/object/object.go]: Formatted.

**Comment by oxddr** [comment]
> this misses an explanation of the new metrics in the README.md

Explanation added.

File: plugin/kubernetes/README.md [hunk 11]
```diff
@@ -229,3 +229,17 @@ plugin is also enabled:
...
+* `coredns_kubernetes_dns_programming_latency_seconds{service_kind}` - exports the
+[DNS programming latency SLI](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/dns_programming_latency.md).
+  The metrics has the `service_kind` label that identifies the kind of the
+  [kubernetes service](https://kubernetes.io/docs/concepts/services-networking/service).
+  It may take one of the three values:
+    * `cluster_ip`
+    * `headless_with_selector`
+    * `headless_without_selector`
+
+TODO(oxddr): add support for other types - only "headless_with_selector" is supported now
```
Comment (miekg) [inline plugin/kubernetes/README.md]: just remove this TODO for now, or move it down below to a BUGS section.

Comment (oxddr) [reply plugin/kubernetes/README.md]: Added "Bugs" section below. I'd rather make it clear, that not all cases are supported. 

For the record, I'm planning to add a support for them soon. I'd rather have this PR merged so we can integrate Kubernetes testing pipeline in parallel.

**Review by miekg — COMMENTED** [review]
no worries. I'll also propose some small changes to have things more inline with other READMEs. But let's merge this first.

**Comment by oxddr** [comment]
Thank you all for the review!
